socorro/external/es/super_search_fields.py

```diff
@@ -176,39 +171,71 @@ def update_field(self, **kwargs):
+            )
+        ):
+            # This is a change that will have an impact on the Elasticsearch
+            # mapping, we first need to make sure it doesn't break.
+            new_mapping = self.get_mapping(overwrite_mapping=params)
+
+            # Try the mapping. If there is an error, an exception will be
+            # raised. If an exception is raised, the new mapping will be
+            # rejected.
+            self.test_mapping(new_mapping)
+
+        if (
+            'storage_mapping' in params
+            and params['storage_mapping'] != old_value['storage_mapping']
+        ):
             # The storage mapping is an object, and thus is treated
             # differently than other fields by Elasticsearch. If a user
             # changes the object by removing a field from it, that field won't
             # be removed as part of the update (which performs a merge of all
-            # objects in the back-end). We therefore want to remove that field
-            # before we do the merge, so that it is entirely overwritten.
-            es_connection.update(
-                index=self.config.elasticsearch.elasticsearch_default_index,
-                doc_type='supersearch_fields',
-                body={'script': 'ctx._source.remove("storage_mapping")'},
-                id=params['name'],
+            # objects in the back-end). We therefore want to perform the merge
+            # ourselves, and remove the field from the database before
+            # re-indexing it.
+            new_doc = old_value.copy()
+            new_doc.update(params)
+
+            es_connection.delete(
+                index=es_index,
+                doc_type=es_doc_type,
+                id=new_doc['name'],
+            )
+            es_connection.index(
+                index=es_index,
+                doc_type=es_doc_type,
+                body=new_doc,
+                id=new_doc['name'],
+                op_type='create',
+                refresh=True,
             )
 
-        # Then update the new field in the database. Note that Elasticsearch
-        # takes care of merging the new document into the old one, so missing
-        # values won't be changed.
-        es_connection.update(
-            index=self.config.elasticsearch.elasticsearch_default_index,
-            doc_type='supersearch_fields',
-            body={'doc': params},
-            id=params['name'],
-            refresh=True,
-        )
-
-        if 'storage_mapping' in params:
             # If we made a change to the storage_mapping, log that change.
             self.config.logger.info(
-                'elasticsearch mapping changed for field "%s", '
+                'Elasticsearch mapping changed for field "%s", '
                 'was "%s", now "%s"',
                 params['name'],
                 old_value['storage_mapping'],
-                params['storage_mapping'],
+                new_doc['storage_mapping'],
+            )
+        else:
```

Comment:
Because we want to run that ``update`` every time the storage mapping was not changed,  not just when ``in_database_name`` was changed. 

Basically, what this does is, make a special case when the storage mapping is changed where we do the "merge" of the old and new documents manually, because we have a very specific use case. In every other case, the ``update`` method just does the job perfectly.